### ISSUE TITLE
Exception when code running for long.

### ISSUE THREAD

**Issue by santiago-vanoli** [aws/aws-iot-device-sdk-python#80]
Hi, 

I have a piece of code that reads some data from sensors  and upload that to AWS. It  works without any trouble for some time (at least 6 hours) but it ends crushing with the following error:

```
Wed Oct 18 09:34:55 2017: Exception in thread Thread-12798:
Wed Oct 18 09:34:55 2017: Traceback (most recent call last):
Wed Oct 18 09:34:55 2017:   File "/usr/local/lib/python2.7/threading.py", line 801, in __bootstrap_inner
Wed Oct 18 09:34:55 2017:     self.run()
Wed Oct 18 09:34:55 2017:   File "/usr/local/lib/python2.7/threading.py", line 1073, in run
Wed Oct 18 09:34:55 2017:     self.function(*self.args, **self.kwargs)
Wed Oct 18 09:34:55 2017:   File "/usr/local/lib/python2.7/site-packages/AWSIoTPythonSDK/core/shadow/deviceShadow.py", line 187, in _timerHandler
Wed Oct 18 09:34:55 2017:     del self._tokenPool[srcToken]
Wed Oct 18 09:34:55 2017: KeyError: 'c52a03a1-f677-49de-9d01-858cbcac9fb8'
```

I have no clue on how to determine what's happening and why.
Feel free to tell me if I'm missing information. Any help/suggestion will be appreciated and welcomed

Best Regards, 
Santiago

**Comment by hedeshianaws** [comment]
Thank you for using AWS IoT. Can you confirm which version of the SDK you are using? The line numbers in deviceShadow.py don't seem to be matching v1.2.0.

**Comment by santiago-vanoli** [comment]
Hi, thanks for the quick response. 
"CHANGELOG.rst" goes up tu v1.1.2 so I'll assume that's my version. Should I update to v1.2.0? If so, Should I expect any compatibility issues?

**Comment by hedeshianaws** [comment]
Can you try this workaround for now:

AWSIoTPythonSDK/core/shadow/deviceShadow.py:

From:
```
    def _timerHandler(self, srcActionName, srcToken):
        with self._dataStructureLock:
            # Remove the token
            del self._tokenPool[srcToken]
```
to:
```
    def _timerHandler(self, srcActionName, srcToken):
        with self._dataStructureLock:
            # Don't crash if we try to remove an unknown token
            if srcToken not in self._tokenPool:
                self._logger.warn('Tried to remove non-existent token from pool: %s' % str(srcToken))
                return
            # Remove the token
            del self._tokenPool[srcToken]
```

**Comment by hedeshianaws** [comment]
In addition, can you instrument your code on line 139.5 to print out the srcToken right before it is cancelled and deleted?

https://github.com/aws/aws-iot-device-sdk-python/blob/v1.1.2/AWSIoTPythonSDK/core/shadow/deviceShadow.py#L140

**Comment by santiago-vanoli** [comment]
The workaround did the trick, thanks for the help!!

**Comment by hedeshianaws** [comment]
@santiago-vanoli Thanks for the confirmation. I'm circulating this fix internally for review, and unless we find issues with it, it will be in the next point release.

**Comment by liuszeng** [comment]
Addressed in v1.3.0.
